### DISCUSSION TITLE
Launch jscad from local file link

### DISCUSSION THREAD

**Discussion by EdHubbell** [jscad/OpenJSCAD.org#895] [Q&A]
I have an application that I could use to generate .js files for users to view. I'd like to be able to launch a link so that my application opens the newly created .js file and opens the jscad file for viewing in the browser. 

Ideally I'd like to get to using local file links, like 
http://openjscad.xyz/#file://C:/Users/ed/Downloads/JSCADExample/index.js
or 
http://openjscad.xyz/#file://C:/Users/ed/Downloads/JSCADExample/

I'd be fine with uploading the file to an http link, like 
https://xdblob.z13.web.core.windows.net/ClickOnce/XData/index.js

But I can't even get that format to work. It seems strange that I can easily get stl files to open directly (e.g. https://openjscad.xyz/#http://www.thingiverse.com/download:164128), but I'm having no success with .js files. 

(Note: I can open the files using the Load a JSCAD button on the page, but that's a step beyond what I want for the users).

It may be that I'd have to install our own local version of the website on our servers instead of banging on openjscad.xyz all the time - I'm cool with that. I didn't want to go too far before determining if the capability I wanted was available. 

I'm guessing I'm doing something wrong - Any advice is appreciated. 
~Ed

**Comment by z3dev** [comment]
@EdHubbell welcome!

There’s a little discussion at the User Group which should provide some hints on how to share designs using URI parameters.

https://openjscad.nodebb.com/topic/234/thingiverse

In general, browsers will not allow access to local files. There are all kinds of rules built into the browsers, and it gets complicated.

But, you should be able to share any file that’s available on-line. You could put the file almost anywhere, including your own website.

**Reply by EdHubbell** [reply]
Thanks for the help! I'm not able to get it to work, but I think it must be something to do with Azure blob storage (or how I'm exposing the file over html). I took the example .js from the thingiverse discussion and put it up on our server here: 

https://xdblob.z13.web.core.windows.net/ClickOnce/XData/parametric_butt_hinge_3.7.js
e.g.
https://www.openjscad.xyz/?uri=https://xdblob.z13.web.core.windows.net/ClickOnce/XData/parametric_butt_hinge_3.7.js

Initially it was set for content type text/javascript, which resulted in this error from https://www.openjscad.xyz/ - 

ERROR: Uncaught SyntaxError: Unexpected identifier Line: 23

I tried changing the content type to application/octet-stream, changing ContentDisposition to attachment; filename="parametric_butt_hinge_3.7.js", keeping the new ContentDisposition and moving back to text/javascript - No dice. Same error. My thinking was that maybe it needed to be a download link instead of a file link. 

Then I tried the same steps with a google drive link. Google Drive gets pretty picky with .js files (wants to virus scan them for you), so no dice there. Then tried renaming to .jscad, which still gets me nowhere using 

https://drive.google.com/uc?export=download&id=1xF9gEs4XIHP8_D45drkd1gGzERNgEe5w
e.g. 
https://www.openjscad.xyz/?uri=https://drive.google.com/uc?export=download&id=1xF9gEs4XIHP8_D45drkd1gGzERNgEe5w

With a new error - ERROR: fetch error: 500 Internal Server Error Stack: Error: fetch error: 500 Internal Server Error at https://www.openjscad.xyz/dist/jscad-web.min.js:3201:1532

My guess is that file sharing via Azure Blob Storage or Google Drive is just different enough to not work. A coworker had no problems with a file from Github. Sure enough, I put the file up there, and it works just fine. 

https://raw.githubusercontent.com/EdHubbell/JSACDFileTest/main/parametric_butt_hinge_3.7.js
e.g.
https://www.openjscad.xyz/?uri=https://raw.githubusercontent.com/EdHubbell/JSACDFileTest/main/parametric_butt_hinge_3.7.js

Or maybe there's some allowed server list for the .xyz instance of openjscad that would be different if I deployed it locally. I'm stumped. I'm mostly a desktop and database dev - http protocol intricacies are well outside my wheelhouse.

~Ed

**Reply by hrgdavor** [reply]
@EdHubbell 
I am working on a reusabe worker for integraion of jscad, and was able to create a semi-standalone page

http://3d.hrg.hr/jscad/jscad_standalone.html

use it as a template. just replace the script at the bottom of the page with your jscad code.
The script is not fully standalone(will not work without internet access) as it downloads jscad packages from https://unpkg.com

**Reply by EdHubbell** [reply]
@hrgdavor - Nice! That's a possibility. 

I like it because it takes away hitting a remote server. We can set it up so we mirror any remote jscad packages to domain names that we have control over, so we're not stuck if something gets upgraded. Although I see that unpkg is linked to npm, so I'm guessing those resources would be available for the foreseeable. 

Thanks for the help! I'll give this approach a shot when I get back around to this project in a week or so. Unsuccessful fiddling with Azure blob storage content-disposition attributes is causing me hair loss.

**Reply by hrgdavor** [reply]
This is sth I hacked quickly.

I could fix it so it does not need internet access at all. 
 - a) single standalone file, 
 - b) with js deps in the same folder.

**Reply by z3dev** [reply]
> Thanks for the help! I'm not able to get it to work, but I think it must be something to do with Azure blob storage (or how I'm exposing the file over html). I took the example .js from the thingiverse discussion and put it up on our server here:
> 
> https://xdblob.z13.web.core.windows.net/ClickOnce/XData/parametric_butt_hinge_3.7.js

I get the following error when trying to load this file directly. Which means that the server is configured to block this request.

The requested content does not exist.

HttpStatusCode: 404
ErrorCode: WebContentNotFound
RequestId : 495075b8-001e-0085-140d-846d07000000
TimeStamp : 2021-07-29T00:08:10.5218902Z

> Then I tried the same steps with a google drive link. Google Drive gets pretty picky with .js files (wants to virus scan them for you), so no dice there. Then tried renaming to .jscad, which still gets me nowhere using
> 
> https://drive.google.com/uc?export=download&id=1xF9gEs4XIHP8_D45drkd1gGzERNgEe5w
> e.g.
> https://www.openjscad.xyz/?uri=https://drive.google.com/uc?export=download&id=1xF9gEs4XIHP8_D45drkd1gGzERNgEe5w
> 
> With a new error - ERROR: fetch error: 500 Internal Server Error Stack: Error: fetch error: 500 Internal Server Error at https://www.openjscad.xyz/dist/jscad-web.min.js:3201:1532

This is the error for non-fetchable file types, i.e. those files that do not return known content.

> My guess is that file sharing via Azure Blob Storage or Google Drive is just different enough to not work. A coworker had no problems with a file from Github. Sure enough, I put the file up there, and it works just fine.

GitHub rocks!

In general, if the file contents can be fetched from a server, and those contents look like one of the known types, then the file can be shared via the URI parameter.

The Azure server isn’t configured correctly so the fetch fails.

Google drive is doing something strange and doesn’t return any content directly so the fetch fails. A lot of services redirect requests to other servers, and do it poorly.

Final note: The URI is sent to a back end process that does the down load to a TEMPORARY file. This eliminates any CORS issues (A very nasty web standard) Once the file is downloaded, and the contents are validated, then the TEMPORARY  file is fetched.

**Reply by z3dev** [reply]
> Thanks for the help! I'll give this approach a shot when I get back around to this project in a week or so. Unsuccessful fiddling with Azure blob storage content-disposition attributes is causing me hair loss.

I don't have any hair! The black magic of HTML headers is just too much.

**Comment by hrgdavor** [comment]
there is also https://github.com/danmarshall/jscad-now
